lib/AL_USDMaya/AL/usdmaya/nodes/LayerManager.cpp

```diff
@@ -0,0 +1,437 @@
+  AL_MAYA_CHECK_ERROR(layersArrayHandle.set(builder), errorString);
+  AL_MAYA_CHECK_ERROR(layersArrayHandle.setAllClean(), errorString);
+  AL_MAYA_CHECK_ERROR(dataBlock.setClean(layers()), errorString);
+  return status;
+}
+
+void LayerManager::loadAllLayers()
+{
+  TF_DEBUG(ALUSDMAYA_LAYERS).Msg("LayerManager::loadAllLayers\n");
+  const char* errorString = "LayerManager::loadAllLayers";
+
+  MStatus status;
+  MPlug allLayersPlug = layersPlug();
+  MPlug singleLayerPlug;
+  MPlug idPlug;
+  MPlug anonymousPlug;
+  MPlug serializedPlug;
+  std::string identifierVal;
+  std::string serializedVal;
+  SdfLayerRefPtr layer;
+  // We DON'T want to use evaluate num elements, because we don't want to trigger
+  // a compute - we want the value(s) as read from the file!
+  const unsigned int numElements = allLayersPlug.numElements();
+  for(unsigned int i=0; i < numElements; ++i)
+  {
+    singleLayerPlug = allLayersPlug.elementByPhysicalIndex(i, &status);
+    AL_MAYA_CHECK_ERROR_CONTINUE(status, errorString);
+    idPlug = singleLayerPlug.child(m_identifier, &status);
+    AL_MAYA_CHECK_ERROR_CONTINUE(status, errorString);
+    anonymousPlug = singleLayerPlug.child(m_anonymous, &status);
+    AL_MAYA_CHECK_ERROR_CONTINUE(status, errorString);
+    serializedPlug = singleLayerPlug.child(m_serialized, &status);
+    AL_MAYA_CHECK_ERROR_CONTINUE(status, errorString);
+
+    identifierVal = idPlug.asString(MDGContext::fsNormal, &status).asChar();
+    AL_MAYA_CHECK_ERROR_CONTINUE(status, errorString);
+    if(identifierVal.empty())
+    {
+      MGlobal::displayError(MString("Error - plug ") + idPlug.partialName(true) + "had empty identifier");
+      continue;
+    }
+    serializedVal = serializedPlug.asString(MDGContext::fsNormal, &status).asChar();
+    AL_MAYA_CHECK_ERROR_CONTINUE(status, errorString);
+    if(serializedVal.empty())
+    {
+      MGlobal::displayError(MString("Error - plug ") + serializedPlug.partialName(true) + "had empty serialization");
+      continue;
+    }
+
+    bool isAnon = anonymousPlug.asBool(MDGContext::fsNormal, &status);
+    AL_MAYA_CHECK_ERROR_CONTINUE(status, errorString);
+    if(isAnon)
+    {
+      // Note that the new identifier will not match the old identifier - only the "tag" will be retained
+      layer = SdfLayer::CreateAnonymous(SdfLayer::GetDisplayNameFromIdentifier(identifierVal));
+    }
+    else
+    {
+      SdfLayerHandle layerHandle = SdfLayer::Find(identifierVal);
+      if (layerHandle)
+      {
+        layer = layerHandle;
+      }
+      else
+      {
+        // TODO: currently, there is a small window here, after the find, and before the New, where
+        // another process might sneak in and create a layer with the same identifier, which could cause
+        // an error. This seems unlikely, but we have a discussion with Pixar to find a way to avoid this.
+
+        SdfFileFormatConstPtr fileFormat;
+        if(TfStringStartsWith(serializedVal, "#usda "))
```

Comment:
Not exactly sure about the distinction between the sdf format and the usda file format... I'm guessing that usda  is a superset of sdf, since the sdf format (presumably) can only store the data for a single layer, while usda can serialize a stage.

From a more practical standpoint, I know it can be relatively easy to write out a "sdf" file when you intend to write out "usda", or vice versa - which you get when you call layer.ExportToString() depends on the details of how that layer was initialized.  Thus why I want to support both.

As for usd - usd isn't a file format of it's own, in the strictest sense (ie, I don't believe you'll ever have any files written out to disk that start with a "#usd " tag) - it's just an extension which allows the underlying file content to be either usda or usdc.  Pixar then has a FileFormat to support it, and transparently read the underlying file properly, regardless of which it actually contains.

Lastly - we MIGHT be able to take the bit after the "#", turn it into a token, and feed it in.  However, I wasn't sure this is always kosher... and for now, the only two text-formats we would conceivbly want to support would be Sdf And Usda, so it seemed safest to just hard code those two.  (Also, it would be slower, as string -> token conversion is relatively slow)

Reply:
OK. I did some tests and I couldn't in any case get the fileformat written out besides 'USDA' to my surprise. I'm happy to leave it as is, I havn't read much into the fileformat part of USD so I don't want to comment much more. Thanks for the reply!